# 6.4: Resampling Exercises

## Getting Started

### Import Libraries 

We import our standard libraries and specific objects/libraries at the top level of our notebook.

In [7]:
# Load our previous libraries and objects
import numpy as np
import statsmodels.api as sm
from ISLP import load_data
from ISLP.models import (ModelSpec as MS,
                         summarize,
                         poly)
from sklearn.model_selection import train_test_split

# Import new libraries and objects
from functools import partial
from sklearn.model_selection import \
     (cross_validate,
      KFold,
      ShuffleSplit)
from sklearn.base import clone
from ISLP.models import sklearn_sm

## The Validation Set Approach

We explore the use of the validation set approach in order to estimate the test error rates that result from fitting various linear models on the `Auto` data set.

We use the function `train_test_split()` to split the data into training and validation sets. As there are 392 observations, we split into two equal sets of size 196 using the argument `test_size=196`. We want to set a random seed (`random_state=0`) so that the random results can be reproduced again.

In [2]:
Auto = load_data('Auto')
Auto_train, Auto_valid = train_test_split(Auto,
                                         test_size=196, # split in two
                                         random_state=0) # random seed

Now we can fit a linear regression using only the observations corresponding to the training set `Auto_train`.

In [3]:
hp_mm = MS(['horsepower'])
X_train = hp_mm.fit_transform(Auto_train)
y_train = Auto_train['mpg']
model = sm.OLS(y_train, X_train)
results = model.fit()

We now use the `predict()` method of results evaluated on the model matrix for this model created using the validation data set. We also calculate the validation MSE of our model.

In [4]:
X_valid = hp_mm.transform(Auto_valid)
y_valid = Auto_valid['mpg']
valid_pred = results.predict(X_valid)
np.mean((y_valid - valid_pred)**2)

23.61661706966988

So our estimated test MSE for the linear regression is 23.62.

**What would happen if we used a different set of observations for our training set? Repeat the above
process using a different seed (i.e. set.seed(2)) and report the differences.**

## Leave-One-Out Cross-Validation

The simplest way to cross-validate in Python is to use `sklearn`. The `ISLP` package provides a wrapper (`sklearn_sm()`), that allows us to easily use the cross-validation tools of `sklear`n with models fit by `statsmodels`.

In [5]:
# LOOC is a specific case of cross-validation
# creates as many folds as the cv value
# when k=n then k-fold becomes the same as leave-one-out
#We only know it’s LOOCV because cv = number of rows in your dataset
hp_model = sklearn_sm(sm.OLS,
                      MS(['horsepower']))
X, Y = Auto.drop(columns=['mpg']), Auto['mpg']
cv_results = cross_validate(hp_model,
                            X,
                            Y,
                            cv=Auto.shape[0])
cv_err = np.mean(cv_results['test_score'])
cv_err

24.231513517929216

In [ ]:
# 392 k-folds
Auto.shape[0]

In [ ]:
cv_results

In [ ]:
cv_results['fit_time']

In [ ]:
cv_results['test_score']

In [ ]:
cv_results['test_score'].shape

In [ ]:
cv_results.keys()

The arguments to `cross_validate()` are: an object with the appropriate `fit()`, `predict()`, and `score()` methods, an array of features `X` and a response `Y`. We also included an additional argument `cv` to `cross_validate()`; specifying an integer 
$K$ results in $K$-fold cross-validation.

We can repeat this procedure for increasingly complex polynomial fits. To automate the process, we again use a for loop which iteratively fits polynomial regressions of degree 1 to 5, computes the associated cross-validation error, and stores it in the $i$th element of the vector `cv_error`.

In [6]:
cv_error = np.zeros(5)
H = np.array(Auto['horsepower'])
M = sklearn_sm(sm.OLS)
for i, d in enumerate(range(1,6)):
    X = np.power.outer(H, np.arange(d+1))
    M_CV = cross_validate(M,
                          X,
                          Y,
                          cv=Auto.shape[0])
    cv_error[i] = np.mean(M_CV['test_score'])
cv_error

array([24.23151352, 19.24821312, 19.33498406, 19.42443033, 19.03323073])

**Write a for loop that fits the Auto data using polynomials up to degree 10. Compute the LOOCV
test error rate for each of them. Which degree polynomial fits the data the best according to
the test error rate and the bias-variance trade-off?**

## k-Fold Cross-Validation

Here we use `KFold()` to partition the data into $K=10$ random groups. We use random_state to set a random seed and initialize a vector `cv_error` in which we will store the CV errors corresponding to the polynomial fits of degrees one to five.

In [7]:
cv_error = np.zeros(5)
cv = KFold(n_splits=10,
           shuffle=True,
           random_state=0) # use same splits for each degree
for i, d in enumerate(range(1,6)):
    X = np.power.outer(H, np.arange(d+1))
    M_CV = cross_validate(M,
                          X,
                          Y,
                          cv=cv)
    cv_error[i] = np.mean(M_CV['test_score'])
cv_error

array([24.20766449, 19.18533142, 19.27626666, 19.47848399, 19.13721595])

**Repeat this process using K = 10 and report the results.**

## The Bootstrap

The bootstrap approach can be applied in almost all situations. While there are several implementations of the bootstrap in Python, its use for estimating standard error is simple enough that we write our own function below for the case when our data is stored in a dataframe. We will use the `Portfolio` data set in the `ISLP` package to illustrate.

We will create a function `alpha_func()`, which takes as input a dataframe `D` assumed
to have columns `X` and `Y`, as well as a
vector `idx` indicating which observations should be used to
estimate 
$\alpha$. The function then outputs the estimate for $\alpha$ based on
the selected observations.

In [8]:
Portfolio = load_data('Portfolio')  # starting with dataset 'Portfolio' 
# the alpha_func method is a financial measure of return over investment of something
# if X are our stock values, Y is something you want to benchmark against
def alpha_func(D, idx):  # bootstrapping alpha_func paramter
   # 'D[['X','Y']].loc[idx]': This selects rows from D where idx is True (or where idx is an index list).
   # 'np.cov(...)': This calculates the covariance matrix of the selected rows, assuming that the data in 
   # 'X' and 'Y' columns are used as variables.
   cov_ = np.cov(D[['X','Y']].loc[idx], rowvar=False)  
   # The return statement calculates a statistic related to bootstrapping using elements of the covariance matrix.
   return ((cov_[1,1] - cov_[0,1]) /
           (cov_[0,0]+cov_[1,1]-2*cov_[0,1]))

The following command estimates $\alpha$ using all 100 observations.

In [45]:
alpha_func(Portfolio, range(100))

0.57583207459283

The value returned, 0.575. We want to know if this is an accuracte value or if we could get a better estimate of this.

Next we randomly select 100 observations from range(100), with replacement. This is equivalent to constructing a new bootstrap data set and recomputing $\alpha$ based on the new data set.

In [46]:
# rng = np.random.default_rng(0): This initializes a random number generator (rng) with a seed value of 0.
rng = np.random.default_rng(0)
# rng.choice(100, 100, replace=True): This generates an array of 100 random indices between 0 and 99 with 
# replacement. This means that the same index can be picked multiple times in the sample.
# alpha_func(Portfolio, ...): This calls the alpha_func function with the Portfolio dataset and the array 
# of random indices as arguments. The alpha_func function will then calculate the alpha statistic using the 
# selected rows from the Portfolio dataset based on the random indices.
alpha_func(Portfolio,
           rng.choice(100,
                      100,
                      replace=True))

0.6074452469619004

This time our result was 0.607. This appears to be a better estimate of alpha than previously at 0.575

We've done this experiment once but what we can do is automate it. 

This process can be generalized to create a simple function `boot_SE()` for computing the bootstrap standard error for arbitrary functions that take only a data frame as an argument.

In [11]:
import warnings

# This code performs bootstrapping to estimate the distribution of a statistic (func) calculated on a dataset D. 

def boot_SE(func,
            D,
            n=None,
            B=1000,  # run experiment 1000 times
            seed=0):
    # NOTE: Suppress FutureWarning in ISLP.models.columns
    # The warning is related to Series.__getitem__ treating keys as positions, which is deprecated.
    # Since ISLP is an external library that I don't control, and this specific warning does not
    # affect my current usage, I'm suppressing it to keep the output clean and focused on relevant information.
    warnings.filterwarnings(action='ignore', category=FutureWarning, module='ISLP.models.columns', lineno=151)
    
    rng = np.random.default_rng(seed)
    first_, second_ = 0, 0
    n = n or D.shape[0]
    samples = []
    # for _ in range(B):: This loop runs B times (1000), where B is the number of bootstrap iterations. In each iteration, 
    # a bootstrap sample is drawn and the statistic is calculated on that sample.
    for _ in range(B):
        # each time it enters, it takes the indexes and makes replacements over each iteration
        # idx = rng.choice(D.index, n, replace=True): This line generates a random sample of indices (idx) from the 
        # dataset D. n is the size of each bootstrap sample, and replace=True indicates that sampling is done with 
        # replacement, meaning that the same index can be selected multiple times in the sample.
        idx = rng.choice(D.index,
                         n,
                         replace=True)
        # value = func(D, idx): This line calculates the statistic (func) on the bootstrap sample specified by the indices idx.
        value = func(D, idx)
        # first_ += value and second_ += value**2: These lines accumulate the statistic and its square across all bootstrap 
        # samples. These values will be used later to calculate the variance of the statistic.
        first_ += value
        second_ += value**2
        # samples.append(value): This line appends the calculated statistic for the current bootstrap sample to the samples list.
        samples.append(value)

        # After the loop, the function returns two values:
        # 1. samples: This is a list containing the calculated statistic for each bootstrap sample. It provides a distribution 
        # of the statistic under bootstrapping.
        # 2. np.sqrt(second_ / B - (first_ / B)**2): This calculates the standard deviation of the statistic estimate based on 
        # the bootstrap samples. It uses the accumulated values first_ and second_ to calculate the variance of the statistic estimate.
    return samples, np.sqrt(second_ / B - (first_ / B)**2) # formula for standard error; sum of all the values divided by B, squared

Let’s use our function to evaluate the accuracy of our estimate of $\alpha$ using $B = 1,000$ bootstrap replications.

In [14]:
samples, alpha_SE = boot_SE(alpha_func,
                   Portfolio,
                   B=1000,
                   seed=0)
alpha_SE

0.09118176521277607

In [15]:
np.mean(samples)

0.5821442531025649

We got a value of 0.582. that is half way between both our first predictions of 0.575 and 0.607. With a standard error of 0.09.

This is a way better estimate of our alpha parameter. 

If you just run samples, if will give you the results for each sample.

Law of Large Numbers.

The final output shows that the bootstrap estimate for ${\rm SE}(\hat{\alpha})$ is $0.0912$.

The bootstrap approach can be used to assess the variability of the coefficient estimates and predictions from a statistical learning method. Here we use the bootstrap approach in order to assess the variability of the estimates for 
 and 
, the intercept and slope terms for the linear regression model that uses horsepower to predict mpg in the Auto data set

We start by writing a generic function `boot_OLS()` for bootstrapping a regression model that takes a formula to define the corresponding regression. We use the `clone()` function to make a copy of the formula that can be refit to the new dataframe. This means that any derived features such as those defined by `poly()` which will be re-fit on the resampled data frame.

In [49]:
def boot_OLS(model_matrix, response, D, idx):
    D_ = D.loc[idx]
    Y_ = D_[response]
    X_ = clone(model_matrix).fit_transform(D_)
    return sm.OLS(Y_, X_).fit().params

hp_func = partial(boot_OLS, MS(['horsepower']), 'mpg')

# Use the hp_func() function to create bootstrap estimates for the intercept and slope 
rng = np.random.default_rng(0)
np.array([hp_func(Auto,
          rng.choice(392,
                     392,
                     replace=True)) for _ in range(10)])

array([[39.88064456, -0.1567849 ],
       [38.73298691, -0.14699495],
       [38.31734657, -0.14442683],
       [39.91446826, -0.15782234],
       [39.43349349, -0.15072702],
       [40.36629857, -0.15912217],
       [39.62334517, -0.15449117],
       [39.0580588 , -0.14952908],
       [38.66688437, -0.14521037],
       [39.64280792, -0.15555698]])

Next, we use the `boot_SE()` {} function to compute the standard errors of 1,000 bootstrap estimates for the intercept and slope terms.

In [50]:
hp_se = boot_SE(hp_func,
                Auto,
                B=1000,
                seed=10)
hp_se

intercept     0.848807
horsepower    0.007352
dtype: float64

The bootstrap estimate for ${\rm SE}(\hat{\beta}_0)$ is
0.85, and that the bootstrap
estimate for ${\rm SE}(\hat{\beta}_1)$ is
0.0074. Standard formulas can be used to compute the standard errors for the regression coefficients in a linear model. These can be obtained using the `summarize()` function from `ISLP.sm`.

In [51]:
hp_model.fit(Auto, Auto['mpg'])
model_se = summarize(hp_model.results_)['std err']
model_se

/Users/Dan/miniconda3/envs/learners/lib/python3.11/site-packages/ISLP/models/__init__.py:49: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  results_table = pd.read_html(tab.as_html(),


intercept     0.717
horsepower    0.006
Name: std err, dtype: float64

The standard error estimates for $\hat{\beta}_0$ and $\hat{\beta}_1$ are 0.717 for the intercept and 0.006 for the slope.

Now, we want to compute the boostrap:

In [52]:
quad_model = MS([poly('horsepower', 2, raw=True)])
quad_func = partial(boot_OLS,
                    quad_model,
                    'mpg')
boot_SE(quad_func, Auto, B=1000)

intercept                                  2.067840
poly(horsepower, degree=2, raw=True)[0]    0.033019
poly(horsepower, degree=2, raw=True)[1]    0.000120
dtype: float64

We compare the results to the standard errors computed using `sm.OLS()`.

In [53]:
M = sm.OLS(Auto['mpg'],
           quad_model.fit_transform(Auto))
summarize(M.fit())['std err']

/Users/Dan/miniconda3/envs/learners/lib/python3.11/site-packages/ISLP/models/__init__.py:49: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  results_table = pd.read_html(tab.as_html(),


intercept                                  1.800
poly(horsepower, degree=2, raw=True)[0]    0.031
poly(horsepower, degree=2, raw=True)[1]    0.000
Name: std err, dtype: float64

*These exercises were adapted from :* James, Gareth, et al. An Introduction to Statistical Learning: with Applications in Python, Springer, 2023.